src/main/haskell/language-kore/src/Data/Kore/ASTUtils/Substitution.hs

```diff
@@ -0,0 +1,129 @@
+License     : UIUC/NCSA
+Maintainer  : phillip.harris@runtimeverification.com
+Stability   : experimental
+Portability : portable
+-}
+
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE Rank2Types             #-}
+{-# LANGUAGE AllowAmbiguousTypes    #-}
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE BangPatterns           #-}
+{-# LANGUAGE LambdaCase             #-}
+{-# LANGUAGE DeriveFunctor          #-}
+{-# LANGUAGE DeriveFoldable         #-}
+{-# LANGUAGE DeriveTraversable      #-}
+{-# LANGUAGE TypeFamilies           #-}
+{-# LANGUAGE PatternSynonyms        #-}
+{-# LANGUAGE ScopedTypeVariables    #-}
+{-# LANGUAGE TypeApplications       #-}
+{-# OPTIONS_GHC -Wno-unused-matches #-}
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+
+module Data.Kore.ASTUtils.Substitution where
+
+
+import           Control.Lens
+
+import qualified Data.Set as S
+import           Data.Fix
+import           Data.List
+
+import           Data.Kore.AST.Common
+import           Data.Kore.AST.MetaOrObject
+import           Data.Kore.AST.PureML
+
+import           Data.Kore.ASTUtils.SmartConstructors
+
+
+subst
+  :: MetaOrObject level 
+  => CommonPurePattern level 
+  -> CommonPurePattern level 
+  -> CommonPurePattern level 
+  -> CommonPurePattern level
+subst a b = \case 
+  Forall_ s1 v p -> handleBinder a b Forall_ s1 v p 
+  Exists_ s1 v p -> handleBinder a b Exists_ s1 v p 
+  pat 
+   | pat == a  -> b 
+   | otherwise -> Fix $ fmap (subst a b) $ unFix pat 
+
+-- subst2 a b = cata (\pat -> if Fix pat == a then b else Fix pat)
+
+handleBinder
+  :: MetaOrObject level 
+  => CommonPurePattern level 
+  -> CommonPurePattern level 
+  -> (Sort level -> Variable level -> CommonPurePattern level -> CommonPurePattern level)
+  -> Sort level 
+  -> Variable level 
+  -> CommonPurePattern level 
+  -> CommonPurePattern level 
+handleBinder a b binder s1 v p = 
+    let fa = freeVars a 
+        fb = freeVars b
+    in if S.member v fa 
+        then binder s1 v p 
+    else if S.member v fb
+        then subst a b $ alphaRename $ binder s1 v p 
+        else binder s1 v $ subst a b p
+
+freeVars
+  :: MetaOrObject level 
+  => CommonPurePattern level
+  -> S.Set (Variable level)
+freeVars (Forall_ s1 v p) = S.delete v $ freeVars p 
+freeVars (Exists_ s1 v p) = S.delete v $ freeVars p 
+freeVars (Var_ v) = S.singleton v 
+freeVars p = S.unions $ map freeVars $ p ^. partsOf allChildren
+
+alphaRename
+  :: MetaOrObject level 
+  => CommonPurePattern level 
+  -> CommonPurePattern level 
+alphaRename = (\case 
+ Forall_ s1 v p -> go Forall_ s1 v p 
+ Exists_ s1 v p -> go Exists_ s1 v p
+ -- x -> x
+ _ -> error "Input must be an Exist or Forall at the top level"
+ ) where 
+    go binder s1 v p = 
+        binder s1 (replacementVar v p)
+        (subst (Var_ v) (Var_ $ replacementVar v p) p)
+    replacementVar v p = head $ alternatives v \\ (S.toList $ freeVars p)
```

Comment:
Isn't it O(m*n) right now, where m is the first index that was not already used, and n is the free var count?

Even if my estimate is correct, I guess it may be reasonable if m is likely to be small (which, by default, it should be). I can't tell right now if it would increase when doing many similar substitutions as we execute code and produce proofs.

Reply:
Whoops, I thought \\ assumes sorted. Yeah, it is slower in this case.

Reply:
But I just noticed: `freeVars p` is being called over and over in the "filter" case. That's another accident that can't happen when you use operations over whole datastructures like union/intersect/difference.

Reply:
Thanks. As I said, I was just curious. We're not supposed to optimize things right now, and we should prefer code that is safer/easier to read/whatever.

Reply:
Yeah I'm just explaining why I think it's a good default. For now I filtered with S.member, in the future maybe substitution could be outsourced to kmett's bound or something.